File: src/org/codepulsar/pulsar/Lexer.java
package org.codepulsar.pulsar;

import java.util.ArrayList;
import static org.codepulsar.pulsar.TokenType.*;

public class Lexer {
    private final String sourceCode;
    private int start;
    private int current;
    private int line;

    private final ArrayList<Error> errors;
    public boolean hasErrors;

    public Lexer(String sourceCode) {
        this.sourceCode = sourceCode;
        this.start = 0;
        this.current = 0;
        this.line = 1;

        this.errors = new ArrayList<>();
        this.hasErrors = false;
    }

    public ArrayList<Token> tokenize() {
        ArrayList<Token> tokens = new ArrayList<>();

        while (true) {
            Token toAdd = scanToken();
            tokens.add(toAdd);

            if (toAdd.getTokenType() == TK_EOF) {
                break;
            }
        }

        return tokens;
    }

    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
    }

    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    private Token scanToken() {
        skipWhitespace();
        this.start = this.current;

        if (isAtEnd()) {
            return makeToken(TK_EOF);
        }

        char now = advance();

        if (isAlpha(now)) {
            return identifier();
        } if (isDigit(now)) {
            return number();
        }

        switch (now) {
            case '.':
                return makeToken(TK_DOT);
            case ',':
                return makeToken(TK_COMMA);
            case ':':
                if (peek() == ':') {
                    advance();
                    return makeToken(TK_DOUBLE_COLON);
                } else {
                    return makeToken(TK_COLON);
                }
            case ';':
                return makeToken(TK_SEMICOLON);
            case '?':
                return makeToken(TK_QMARK);
            case '(':
                return makeToken(TK_LPAR);
            case ')':
                return makeToken(TK_RPAR);
            case '{':
                return makeToken(TK_LBRACE);
            case '}':
                return makeToken(TK_RBRACE);
            case '"':
                return string();

            case '!':
                if (peek() == '=') {
                    advance();
                    return makeToken(TK_NOT_EQUAL);
                } else {
                    return makeToken(TK_NOT);
                }
            case '+':
                if (peek() == '='){
                    advance();
                    return makeToken(TK_PLUS_EQUAL);
                } else {
                    return makeToken(TK_PLUS);
                }
            case '-':if (peek() == '='){
                    advance();
                    return makeToken(TK_MINUS_EQUAL);
                } else {
                    return makeToken(TK_MINUS);
                }
            case '*':
                if (peek() == '=') {
                    advance();
                    return makeToken(TK_MUL_EQUAL);
                } else {
                    return makeToken(TK_MULTIPLICATION);
                }
            case '/':
                if (peek() == '=') {
                    advance();
                    return makeToken(TK_DIV_EQUAL);
                } else {
                    return makeToken(TK_DIVISION);
                }
            case '%':
                if (peek() == '=') {
                    advance();
                    return makeToken(TK_MOD_EQUAL);
                } else {
                    return makeToken(TK_MODULUS);
                }

            case '=':
                if (peek() == '=') {
                    advance();
                    return makeToken(TK_EQUAL_EQUAL);
                } else {
                    return makeToken(TK_EQUAL);
                }
            case '>':
                if (peek() == '=') {
                    advance();
                    return makeToken(TK_GT_EQUAL);
                } else {
                    return makeToken(TK_GT);
                }
            case '<':
                if (peek() == '=') {
                    advance();
                    return makeToken(TK_LT_EQUAL);
                } else {
                    return makeToken(TK_LT);
                }
            case '|':
                if (peek() == '|') {
                    advance();
                    return makeToken(TK_LOGICAL_OR);
                } else {
                    return errorToken("Invalid Character: '" + now + "'. Perhaps You Meant Logical Or: ||");
                }
            case '&':
                if (peek() == '&') {
                    advance();
                    return makeToken(TK_LOGICAL_AND);
                } else {
                    return errorToken("Invalid Character: '" + now + "'. Perhaps You Meant Logical AND: &&");
                }
        }

        return errorToken("Invalid Character: " + now);
    }

    private Token string() {
        while (peek() != '"' && !isAtEnd()) {
            if (peek() == '\n') {
                this.line++;
            }

            advance();
        }

        if (isAtEnd()) {
            return errorToken("String Literal Not Closed");
        }

        advance();
        return makeToken(TK_STRING);
    }

    private Token identifier() {
        while (isAlpha(peek()) || isDigit(peek()))  {
            advance();
        }

        return makeToken(identifyIdentifier());
    }

    private TokenType identifyIdentifier() {
        return switch (this.sourceCode.charAt(this.start)) {
            case 'a' -> switch (this.sourceCode.charAt(this.start + 1)) {
                case 's' -> checkKeyword("ync", TK_ASYNC, 2);
                case 'w' -> checkKeyword("ait", TK_AWAIT, 2);
                default -> TK_IDENTIFIER;
            };
            case 'b' -> checkKeyword("reak", TK_BREAK, 1);
            case 'c' -> switch (this.sourceCode.charAt(this.start + 1)) {
                case 'l' -> checkKeyword("ass", TK_CLASS, 2);
                case 'o' -> switch (this.sourceCode.charAt(this.start + 3)) {
                    case 's' -> checkKeyword("t", TK_CONST, 4);
                    case 't' -> checkKeyword("inue", TK_CONTINUE, 4);
                    default -> TK_IDENTIFIER;
                };
                default -> TK_IDENTIFIER;
            };
            case 'e' -> switch (this.sourceCode.charAt(this.start + 1)) {
                case 'l' -> checkKeyword("se", TK_ELSE, 2);
                case 'n' -> checkKeyword("um", TK_ENUM, 2);
                case 'x' -> checkKeyword("cept", TK_EXCEPT, 2);
                default -> TK_IDENTIFIER;
            };
            case 'f' -> switch (this.sourceCode.charAt(this.start + 1)) {
                case 'u' -> checkKeyword("n", TK_FUN, 2);
                case 'a' -> checkKeyword("lse", TK_FALSE, 2);
                default -> TK_IDENTIFIER;
            };
            case 'i' -> switch (this.sourceCode.charAt(this.start + 1)) {
                case 'f' -> checkKeyword("", TK_IF, 2);
                case 'm' -> checkKeyword("port", TK_IMPORT, 2);
                default -> TK_IDENTIFIER;
            };
            case 'l' -> checkKeyword("oop", TK_LOOP, 1);
            case 'm' -> switch (this.sourceCode.charAt(this.start + 1)) {
                case 'o' -> checkKeyword("d", TK_MOD, 2);
                case 'a' -> checkKeyword("tch", TK_MATCH, 2);
                default -> TK_IDENTIFIER;
            };
            case 'n' -> switch (this.sourceCode.charAt(this.start + 1)) {
                case 'u' -> checkKeyword("ll", TK_NULL, 2);
                case 'a' -> checkKeyword("tive", TK_NATIVE, 2);
                default -> TK_IDENTIFIER;
            };
            case 'p' -> switch (this.sourceCode.charAt(this.start + 1)) {
                case 'a' -> checkKeyword("ckage", TK_PACKAGE, 2);
                case 'r' -> checkKeyword("int", TK_PRINT, 2);
                default -> TK_IDENTIFIER;
            };
            case 'r' -> checkKeyword("eturn", TK_RETURN, 1);
            case 't' -> switch (this.sourceCode.charAt(this.start + 2)) {
                case 'u' -> checkKeyword("e", TK_TRUE, 3);
                case 'y' -> checkKeyword("", TK_TRY, 3);
                default -> TK_IDENTIFIER;
            };
            case 'u' -> checkKeyword("nsafe", TK_UNSAFE, 1);
            case 'v' -> checkKeyword("ar", TK_VAR, 1);
            case 'w' -> checkKeyword("hile", TK_WHILE, 1);
            default -> TK_IDENTIFIER;
        };
    }

    private TokenType checkKeyword(String remaining, TokenType tokenType, int starter) {
        int length = remaining.length() + starter;
        String word = this.sourceCode.substring(start + starter, start + length);

        if (((this.current - this.start) == length) && (word.equals(remaining))) {
            return tokenType;
        }

        return TK_IDENTIFIER;
    }

    private Token number() {
        int dotCount = 0;
        while (isDigit(peek()) || peek() == '.') {
            if (peek() == '.') {
                dotCount++;
            }
            advance();
        }

        if (dotCount == 1 && !currentLiteral().endsWith(".")) {
            return makeToken(TK_DOUBLE);
        } else if (dotCount > 1) {
            return errorToken("Invalid Literal: " + currentLiteral());
        } else if (currentLiteral().endsWith(".")) {
            return errorToken("Invalid Literal: " + currentLiteral());
        }

        return makeToken(TK_INTEGER);
    }

    private Token makeToken(TokenType tokenType) {
        return new Token(tokenType, currentLiteral(), this.line);
    }

    private Token errorToken(String message) {
        this.hasErrors = true;

        Token errorToken = new Token(TK_ERROR, message, this.line);
        Error error = new Error("Tokenizing Error", message, errorToken);

        this.errors.add(error);
        return errorToken;
    }

    private void skipWhitespace() {
        while (true) {
            char next = peek();

            switch (next) {
                case ' ', '\r', '\t':
                    advance();
                    break;
                case '\n':
                    this.line++;
                    advance();
                    break;

                case '/':
                    if (peek(1) == '/') {
                        while (peek() != '\n' && !isAtEnd()) {
                            advance();
                        }
                    } else {
                        return;
                    }
                    break;

                default:
                    return;
            }
        }
    }

    public ArrayList<Error> getErrors() {
        return this.errors;
    }

    private String currentLiteral() {
        return this.sourceCode.substring(this.start, this.current);
    }

    private boolean isAtEnd() {
        return this.current >= this.sourceCode.length() - 1;
    }

    private char peek() {
        if (isAtEnd()) {
            return '\0';
        }
        return this.sourceCode.charAt(this.current);
    }

    private char peek(int skip) {
        if (isAtEnd()) {
            return '\0';
        }
        return this.sourceCode.charAt(this.current + skip);
    }

    private char advance() {
        this.current++;
        return this.sourceCode.charAt(this.current - 1);
    }
}

File: src/codepulsar/pulsar/Lexer.cpp
#include "Lexer.h"


Pulsar::Lexer::Lexer(std::string sourceCode) {
    this->sourceCode = sourceCode;

    this->start = 0;
    this->current = 0;
    this->line = 1;

    this->errors = new CompilerError();
}

std::vector<Pulsar::Token> Pulsar::Lexer::tokenize() {
    if (this->sourceCode.empty()) return this->tokens;

    while (true) {
        Token toAdd = scanToken();
        this->tokens.push_back(toAdd);

        if (toAdd.tokenType == TK_EOF) break;
    }

    return this->tokens;
}

Pulsar::Token Pulsar::Lexer::scanToken() {
    skipWhitespace();
    this->start = this->current;

    if (isAtEnd()) return makeToken(TK_EOF);

    char now = advance();
    if (isAlpha(now)) { return scanIdentifier(); }
    if (isDigit(now)) { return scanNumber(); }
    return scanCharacter(now);
}

Pulsar::Token Pulsar::Lexer::scanIdentifier() {
    while (isAlpha(peek()) || isDigit(peek())) {
        advance();
    }

    return makeToken(identifyIdentifier());
}

Pulsar::TokenType Pulsar::Lexer::identifyIdentifier() {
    if (currentLiteral() == "async") return TK_ASYNC;
    if (currentLiteral() == "await") return TK_AWAIT;
    if (currentLiteral() == "break") return TK_BREAK;
    if (currentLiteral() == "boolean") return TK_BOOLEAN_TYPE;
    if (currentLiteral() == "char") return TK_CHAR_TYPE;
    if (currentLiteral() == "class") return TK_CLASS;
    if (currentLiteral() == "const") return TK_CONST;
    if (currentLiteral() == "continue") return TK_CONTINUE;
    if (currentLiteral() == "double") return TK_DOUBLE_TYPE;
    if (currentLiteral() == "else") return TK_ELSE;
    if (currentLiteral() == "enum") return TK_ENUM;
    if (currentLiteral() == "except") return TK_EXCEPT;
    if (currentLiteral() == "extends") return TK_EXTENDS;
    if (currentLiteral() == "fun") return TK_FUN;
    if (currentLiteral() == "false") return TK_FALSE;
    if (currentLiteral() == "if") return TK_IF;
    if (currentLiteral() == "implements") return TK_IMPLEMENTS;
    if (currentLiteral() == "import") return TK_IMPORT;
    if (currentLiteral() == "init") return TK_INIT;
    if (currentLiteral() == "int") return TK_INT_TYPE;
    if (currentLiteral() == "internal") return TK_INTERNAL;
    if (currentLiteral() == "loop") return TK_LOOP;
    if (currentLiteral() == "match") return TK_MATCH;
    if (currentLiteral() == "method") return TK_METHOD;
    if (currentLiteral() == "mod") return TK_MOD;
    if (currentLiteral() == "namespace") return TK_NAMESPACE;
    if (currentLiteral() == "native") return TK_NATIVE;
    if (currentLiteral() == "override") return TK_OVERRIDE;
    if (currentLiteral() == "package") return TK_PACKAGE;
    if (currentLiteral() == "print") return TK_PRINT;
    if (currentLiteral() == "private") return TK_PRIVATE;
    if (currentLiteral() == "public") return TK_PUBLIC;
    if (currentLiteral() == "return") return TK_RETURN;
    if (currentLiteral() == "static") return TK_STATIC;
    if (currentLiteral() == "trait") return TK_TRAIT;
    if (currentLiteral() == "true") return TK_TRUE;
    if (currentLiteral() == "try") return TK_TRY;
    if (currentLiteral() == "unsafe") return TK_UNSAFE;
    if (currentLiteral() == "use") return TK_USE;
    if (currentLiteral() == "var") return TK_VAR;
    if (currentLiteral() == "void") return TK_VOID;
    if (currentLiteral() == "while") return TK_WHILE;
    else return TK_IDENTIFIER;
}

Pulsar::Token Pulsar::Lexer::scanNumber() {
    int dotCount = 0;

    while (isDigit(peek()) || peek() == '.') {
        if (peek() == '.') dotCount++;
        advance();
    }

    if (dotCount == 1 && currentLiteral().back() != '.') {
        return makeToken(TK_DOUBLE);
    } else if (dotCount > 1) {
        return errorToken("Invalid Number Literal: " + currentLiteral());
    } else if (currentLiteral().back() == '.') {
        return errorToken("Invalid Number Literal: " + currentLiteral());
    }

    try {
        stoi(currentLiteral());
    } catch (std::exception) {
        errorToken("Invalid Value For Integer: " + currentLiteral());
    }

    return makeToken(TK_INTEGER);
}

Pulsar::Token Pulsar::Lexer::scanCharacter(char character) {
    switch (character) {
        case '.':
            return makeToken(TK_DOT);
        case ',':
            return makeToken(TK_COMMA);
        case ':':
            if (peek() == ':') {
                advance();
                return makeToken(TK_DOUBLE_COLON);
            } else {
                return makeToken(TK_COLON);
            }
        case ';':
            return makeToken(TK_SEMICOLON);
        case '?':
            return makeToken(TK_QMARK);
        case '(':
            return makeToken(TK_LPAR);
        case ')':
            return makeToken(TK_RPAR);
        case '{':
            return makeToken(TK_LBRACE);
        case '}':
            return makeToken(TK_RBRACE);
        case '\'':
            return charLiteral();

        case '!':
            if (peek() == '=') {
                advance();
                return makeToken(TK_NOT_EQUAL);
            } else {
                return makeToken(TK_NOT);
            }
        case '+':
            if (peek() == '=') {
                advance();
                return makeToken(TK_PLUS_EQUAL);
            } else {
                return makeToken(TK_PLUS);
            }
        case '-':
            if (peek() == '=') {
                advance();
                return makeToken(TK_MINUS_EQUAL);
            } else if (peek() == '>') {
                advance();
                return makeToken(TK_ARROW);
            } else {
                return makeToken(TK_MINUS);
            }
        case '*':
            if (peek() == '=') {
                advance();
                return makeToken(TK_MUL_EQUAL);
            } else {
                return makeToken(TK_MULTIPLICATION);
            }
        case '/':
            if (peek() == '=') {
                advance();
                return makeToken(TK_DIV_EQUAL);
            } else {
                return makeToken(TK_DIVISION);
            }
        case '%':
            if (peek() == '=') {
                advance();
                return makeToken(TK_MOD_EQUAL);
            } else {
                return makeToken(TK_MODULUS);
            }

        case '=':
            if (peek() == '=') {
                advance();
                return makeToken(TK_EQUAL_EQUAL);
            } else {
                return makeToken(TK_EQUAL);
            }
        case '>':
            if (peek() == '=') {
                advance();
                return makeToken(TK_GT_EQUAL);
            } else {
                return makeToken(TK_GT);
            }
        case '<':
            if (peek() == '=') {
                advance();
                return makeToken(TK_LT_EQUAL);
            } else {
                return makeToken(TK_LT);
            }
        case '|':
            if (peek() == '|') {
                advance();
                return makeToken(TK_LOGICAL_OR);
            } else {
                return errorToken(std::string("Invalid Character: '") + character + std::string("'. Perhaps You Meant Logical Or: ||"));
            }
        case '&':
            if (peek() == '&') {
                advance();
                return makeToken(TK_LOGICAL_AND);
            } else {
                return errorToken(std::string("Invalid Character: '") + character + std::string("'. Perhaps You Meant Logical And: &&"));
            }

        default: return errorToken(std::string("Invalid Character: ") + character);
    }
}

Pulsar::Token Pulsar::Lexer::charLiteral() {
    while (peek() != '\'' && !isAtEnd()) {
        if (peek() == '\n') this->line++;
        advance();
    }

    if (isAtEnd()) return errorToken("The Char Literal Has Not Been Closed");
    advance();
    if (currentLiteral().length() != 3) return errorToken("A Char May Only Contain 1 Character");

    return makeToken(TK_CHARACTER);
}

Pulsar::Token Pulsar::Lexer::makeToken(Pulsar::TokenType tokenType) {
    return Pulsar::Token(tokenType, currentLiteral(), this->line);
}

Pulsar::Token Pulsar::Lexer::errorToken(std::string message) {
    Pulsar::Token errorToken = Token(TK_ERROR, message, this->line);
    this->errors->addError("Tokenizing Error", message, this->line);
    return errorToken;
}

void Pulsar::Lexer::skipWhitespace() {
    while (true) {
        char next = peek();

        switch (next) {
            case ' ': case '\r': case '\t':
                advance();
                break;

            case '\n':
                this->line++;
                advance();
                break;

            case '/':
                if (peek(1) == '/') { while (peek() != '\n' && !isAtEnd()) { advance(); } }
                else return;
                break;

            default:
                return;
        }
    }
}

std::string Pulsar::Lexer::currentLiteral() {
    return this->sourceCode.substr(this->start, this->current - this->start);
}

bool Pulsar::Lexer::isAlpha(char c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
}

bool Pulsar::Lexer::isDigit(char c) {
    return c >= '0' && c <= '9';
}

char Pulsar::Lexer::peek() {
    if (isAtEnd()) return '\0';
    return this->sourceCode[this->current];
}

char Pulsar::Lexer::peek(int skip) {
    if (isAtEnd()) return '\0';
    return this->sourceCode[this->current + skip];
}

char Pulsar::Lexer::advance() {
    this->current++;
    return this->sourceCode[this->current - 1];
}

bool Pulsar::Lexer::isAtEnd() {
    return this->current >= this->sourceCode.length();
}

Pulsar::CompilerError* Pulsar::Lexer::getErrors() {
    return this->errors;
}